rerun_py/rerun_sdk/rerun/_image.py

```diff
@@ -14,28 +13,49 @@
 from .datatypes import TensorBuffer, TensorDimension
 
 
-class ImageFormat(Enum):
+class ImageFormat:
     """Image file format."""
 
-    BMP = "BMP"
-    """BMP format."""
+    name: str
 
-    GIF = "GIF"
-    """GIF format."""
+    BMP: ImageFormat
+    GIF: ImageFormat
+    JPEG: ImageFormat
+    PNG: ImageFormat
+    TIFF: ImageFormat
+    NV12: type[NV12]
 
-    JPEG = "JPEG"
-    """JPEG format."""
-
-    PNG = "PNG"
-    """PNG format."""
-
-    TIFF = "TIFF"
-    """TIFF format."""
+    def __init__(self, name: str):
+        self.name = name
 
     def __str__(self) -> str:
         return self.name
 
 
+class NV12(ImageFormat):
+    """NV12 format."""
+
+    name = "NV12"
+
+    def __init__(self, width: int | None = None, height: int | None = None) -> None:
```

Comment:
We should not put size information into the image format. Instead the expected shape for NV12 tensors should be documented. To allow the convenient automatic reshaping as down below in response to NV12 we can instead add a size hint parameter to a custom __init__ method of Tensor and do the necessary reshaping & validation there in response. Let me know if you want help/clarification with this!

Reply:
@Wumpf If I understand correctly this would look something like this:

In the ImageEncoded component class the __init__ method would accept the following as contents:
```py
        contents: bytes | IO[bytes] | npt.NDArray | rr.Tensor | None = None,
```
npt.NDArray to allow for np arrays (because bytes don't have shapes) and rr.Tensor to leverage an updated TensorExt __init__ which accepts a size hint to allow for an automatic resize. Now I am not too sure how an automatic resize would look like on a tensor, since it's kind of a general purpose datatype: did you have a specific api in mind?

Reply:
Looking at things a bit closer I realize that the `ImageEncoded` api is a little bit in flux actually (and `log_image_file` which you used in the PR description will be removed on 0.10) and I got the entries of the whole thing wrong. Also I didn't give it much thought that you wanted to pass in just bytes which totally makes sense here.
I'm actually not sure if we want `ImageEncoded` to mean only file formats (status quo) or _also_ anything that isn't rgb (like used here now) in the long run 🤔 ...
Anyways, let's not get into a rabbit hole here and keep it simple, sticking with `ImageEncoded`.

What do you think of this (incomplete diff against main):
```diff
diff --git a/rerun_py/rerun_sdk/rerun/_image.py b/rerun_py/rerun_sdk/rerun/_image.py
index 65ca974eb..85738a992 100644
--- a/rerun_py/rerun_sdk/rerun/_image.py
+++ b/rerun_py/rerun_sdk/rerun/_image.py
@@ -3,11 +3,13 @@ from __future__ import annotations
 import io
 import pathlib
 from enum import Enum
-from typing import IO, Iterable
+from typing import IO, Iterable, Sequence
 
 import numpy as np
 from PIL import Image as PILImage
 
+from rerun.error_utils import _send_warning_or_raise
+
 from ._log import AsComponents, ComponentBatchLike
 from .archetypes import Image
 from .components import DrawOrderLike, TensorData
@@ -53,6 +55,7 @@ class ImageEncoded(AsComponents):
         contents: bytes | IO[bytes] | None = None,
         format: ImageFormat | None = None,
         draw_order: DrawOrderLike | None = None,
+        size_hint: Sequence[int] | None = None,
     ) -> None:
         """
         Create a new image with a given format.
@@ -72,6 +75,10 @@ class ImageEncoded(AsComponents):
             An optional floating point value that specifies the 2D drawing
             order. Objects with higher values are drawn on top of those with
             lower values.
+        size_hint:
+            An optional tuple/list of (width, height) that specifies the image size.
+            This is *required* for formats that do not implicitly specify their size
+            and is ignored for formats that do (all file formats).
         """
         if (path is None) == (contents is None):
             raise ValueError("Must provide exactly one of 'path' or 'contents'")
@@ -89,6 +96,11 @@ class ImageEncoded(AsComponents):
 
         if format is not None:
             formats = (str(format),)
+            if isinstance(format, NV12):
+                if size_hint is None:
+                    raise ValueError("Must provide size_hint for NV12 images")
+            elif size_hint is not None:
+                _send_warning_or_raise(f"size_hint is ignored for images with {format}")
         else:
             formats = None
```

Not so sure about `size_hint` as name. And admittedly this is still rather nasty since now Nv12 is the one and only format that needs it and everything else just ignores the size :/. But I think that's better to have some of the formats embed a size and some not (if we keep on this path there's gonna be more of them that follow the Nv12 pattern).
It's pretty clear that we'll have to refactor this again later, but would good to have some path forward.

Reply:
Yea let's not complicate it too much. Hear me out: I think embedding the size hint along with the color_standard would make a lot of sense here. I think this will be a better UX, since you wouldn't be able to try and log an NV12 image without supplying it's size + I think it's also a nice place for specifying the [color standard](https://github.com/rerun-io/rerun/pull/3541#discussion_r1349967368) used... These are all NV12 specific right now, that's why I'd like to keep these fields hidden from the general ImageEncoded api...

```py
class ColorStandard(Enum):
    Bt601 = "Bt601"
    Bt709 = "Bt709"


class NV12(ImageFormat):
    """NV12 format."""

    name = "NV12"
    size_hint: Tuple[int, int]
    color_standard: ColorStandard

    def __init__(self, size_hint: Tuple[int, int], color_standard: ColorStandard = ColorStandard.Bt601) -> None:
        """
        An NV12 encoded image.

        Parameters
        ----------
        size_hint:
            (height, width), the RGB size of the image
        color_standard:
            rr.ColorStandard (default: rr.ColorStandard.Bt601),
            the color standard used to decode the image.(ref. https://en.wikipedia.org/wiki/YCbCr#ITU-R_BT.601_conversion)
        """
        self.size_hint = size_hint
        self.color_standard: color_standard
 ```
 
 I am biased because I wrote this code, but I really like the way this comes out as an API, as long as you are using a linter.
 
 This allows for:
 ```py
 rr.log("NV12", contents=byte_contents, format=rr.ImageFormat.NV12((1080, 1920), rr.ColorFormat.Bt601))
 ```
 Instead of:
 ```py
 rr.log("NV12", contents=byte_contents, format=rr.ImageFormat.NV12, size_hint, color_format)
 ```
 The latter doesn't look bad for NV12 either, but I think when you are logging let's say a JPEG and see the size_hint and color_format fields they may confuse you...
 
 In addition, I think being able to pass a `np.array` to ImageEncoded in combination with bytes, could also be a very welcome *future* improvement.

Reply:
apologies for the reply delay!
Checked with some other folks on the team as well and concluded you're right, the ergonomics of this are way better :)
One thing thing though, it would be nice to put `ColorStandard` as a separate component on image similar to how DepthMeter is done on depth image. You'd in any case run into problems of how to pipe it through otherwise. (and yes we'd just warn for Bt601 on anything but NV12, since effectively we're assuming BT709/srgb everywhere and don't want to implement the conversion right now)

But I realize that this is another big change on top to throw at you and given how big the pr is already I'd even prefer if you just hardcode the primaries that you need (maybe with a note pointing here) and we'll ship that, adding ColorStandard later :)
Meaning we're landing this file here pretty much as is 👍

Reply:
right, I think this means we're good to go then unless you have something more to add?

on my side I need yet to do a quick run on web and maybe add a "test" to the `test_tensor.py` but otherwise this can go to 0.10 :)

Reply:
Glad to hear that! Let's :shipit:. From my side, I could add a simple example that logs NV12 frames from a webcam using cv2 (frames are converted from BGR -> NV12 on host), as you thought of somewhere in this discussion:
~~~py
#!/usr/bin/env python3
"""
Stream NV12 images from a webcam.

Run:
```sh
pip install -r examples/python/nv12/requirements.txt
python examples/python/nv12/main.py
```
"""
from __future__ import annotations

import argparse

import rerun as rr  # pip install rerun-sdk
import cv2
import numpy as np
import time
from inputimeout import inputimeout, TimeoutOccurred


def bgr2nv12(bgr: np.ndarray) -> np.ndarray:
    yuv = cv2.cvtColor(bgr, cv2.COLOR_BGR2YUV_I420)
    uv_row_cnt = yuv.shape[0] // 3
    uv_plane = np.transpose(yuv[uv_row_cnt * 2 :].reshape(2, -1), [1, 0])
    yuv[uv_row_cnt * 2 :] = uv_plane.reshape(uv_row_cnt, -1)
    return yuv


def main() -> None:
    parser = argparse.ArgumentParser(description="Example of using the Rerun visualizer to display NV12 images.")
    rr.script_add_args(parser)
    # Add timeout args
    parser.add_argument(
        "-t",
        "--timeout",
        type=float,
        default=5,
        help="Timeout in seconds, after which the script will stop streaming frames.",
    )
    args = parser.parse_args()

    rr.script_setup(args, "NV12 image example")

    cap = cv2.VideoCapture(0)
    if not cap.isOpened():
        raise RuntimeError("This example requires a webcam.")
    start_time = time.time()
    print(f"Started streaming NV12 images for {args.timeout} seconds.")
    while start_time + args.timeout > time.time():
        ret, frame = cap.read()
        if not ret:
            time.sleep(0.01)
            continue
        rr.log(
            "NV12",
            rr.ImageEncoded(
                contents=bytes(bgr2nv12(np.array(frame))),
                format=rr.ImageFormat.NV12((frame.shape[0], frame.shape[1])),
            ),
        )
        time.sleep(0.01)
    rr.script_teardown(args)


if __name__ == "__main__":
    main()
~~~

Reply:
Oh, and I'd really aprreciate it if you tested for web, thanks!